### DISCUSSION TITLE
Exception handling

### DISCUSSION THREAD

**Discussion by garumble** [miguelgrinberg/microdot#270] [General]
I routinely get hit with a request that gets an ECONNRESET and eventually I run out of memory.  I'm running on an Pico W with the latest microdot.
The error looks like this:
[C]: 11-21-2024 12:11:48 - Traceback (most recent call last):
[C]: 11-21-2024 12:11:48 - File "microdot.py", line 1334, in handle_request
[C]: 11-21-2024 12:11:48 - File "microdot.py", line 401, in create
[C]: 11-21-2024 12:11:49 - File "microdot.py", line 515, in _safe_readline
[C]: 11-21-2024 12:11:49 - File "uasyncio/stream.py", line , in readline
[C]: 11-21-2024 12:11:49 - OSError: [Errno 104] ECONNRESET

Is there a graceful way this I should be using to handle this?
I tried implementing 
@app.errorhandler(OSError)
async def runtime_error(request, exception):

But it never gets called.

**Comment by miguelgrinberg** [comment]
@garumble this basically means that the client opened a connection but did not send a complete request. The error is being handled for you, and for that reason Microdot does not crash and remains running. The handling logs the exception, because it is unexpected. Do you know what might the client be doing that causes this? What is the client? Browser or something else?

Also to clarify, this happens outside of your request handlers, it happens while a request is being received. The error handlers that you configure in your application are intended to catch errors in the handlers themselves, so they do not apply in this case.

**Reply by garumble** [reply]
I set up Wireshark to monitor the transactions to the device and I don't see any odd HTTP requests.
There are a bunch of ARP requests at about the same time as the exceptions.  Any chance it could be those?

**Reply by garumble** [reply]
Is there a way to retrieve the IP address from the exception?

**Reply by miguelgrinberg** [reply]
You ignored my question. Is the client a browser or some other thing? Do you need to know the IP address because you have more than one client connecting to this device? You can modify the code to print the client IP if that is important.

**Reply by garumble** [reply]
The device is an IOT device on our internal network.  Anybody could connect to it.
When I view the traffic in Wireshark, I only see outbound requests from the device to our server.  I don't see any inbound requests to the device.
Modifying the code is probably my next move.

**Reply by miguelgrinberg** [reply]
> The device is an IOT device on our internal network. Anybody could connect to it.

I think the device you mention here is your Microdot server, correct? I'm asking about the client, not the server. I understand that anybody can connect to your web server but they do so with some sort of client application. My question is specifically about the client that was used when these errors were observed. If it was a browser, then it would be useful to know which one (some browsers have strange behaviors, and this could be one of them). If it was not a browser, then I would need to know more about the client, since it is likely the issue originates there.

**Comment by garumble** [comment]
There is only one system that **should** be connecting to the device.  It is a custom application written in C#.  However, I don't know if that is in fact the application that is making the request that causes the exception.  Normally the application is only listening for requests originating from the device.  It is only initiates a request due to an action by a user.  These exceptions occur randomly throughout the day without any user action having taken place.
